### PR TITLE
 OPENEUROPA-2420: Maintain front page alias with language suffix links.

### PR DESCRIPTION
**Pull request by nagyad**

### CODE REVIEW

File: modules/oe_multilingual_front_page/src/PathProcessorFront.php [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+use Drupal\Core\PathProcessor\OutboundPathProcessorInterface;
+use Drupal\Core\Render\BubbleableMetadata;
+use Symfony\Component\HttpFoundation\Request;
+
+/**
+ * Processes the inbound path by resolving it to the front page if empty.
+ *
+ * @todo - remove ::processOutbound() when we remove UrlGenerator::fromPath().
+ */
+class PathProcessorFront implements OutboundPathProcessorInterface {
```
Comment (hernani) [inline modules/oe_multilingual_front_page/src/PathProcessorFront.php]: Rename to PathProcessorFrontPage

File: modules/oe_multilingual_front_page/src/PathProcessorFront.php [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+namespace Drupal\oe_multilingual_front_page;
+
+use Drupal\Core\Config\ConfigFactoryInterface;
+use Drupal\Core\Path\AliasManager;
+use Drupal\Core\PathProcessor\OutboundPathProcessorInterface;
+use Drupal\Core\Render\BubbleableMetadata;
+use Symfony\Component\HttpFoundation\Request;
+
+/**
+ * Processes the inbound path by resolving it to the front page if empty.
```
Comment (hernani) [inline modules/oe_multilingual_front_page/src/PathProcessorFront.php]: Comment does not make sense.

File: modules/oe_multilingual_front_page/src/PathProcessorFront.php [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+ *
+ * @todo - remove ::processOutbound() when we remove UrlGenerator::fromPath().
```
Comment (hernani) [inline modules/oe_multilingual_front_page/src/PathProcessorFront.php]: Comment does not make sense.

File: modules/oe_multilingual_front_page/README.md [hunk 4]
```diff
@@ -0,0 +1,7 @@
+# OpenEuropa Multilingual Front page
+
+The module ensures that "index" is set as homepage alias.
```
Comment (hernani) [inline modules/oe_multilingual_front_page/README.md]: The module ensures that the homepage alias is alsways maintained when linking to the front page.

File: tests/features/language-negotiation-suffix.feature [hunk 5]
```diff
@@ -53,3 +53,12 @@ Feature: Language negotiation by suffix.
...
 
+  @front-page
+  Scenario: Language switcher should maintain front page alias in the links.
+    Given I am logged in as a user with the "administer languages, access administration pages, view the administration theme, administer site configuration" permission
+    And I am on "/admin/config/system/site-information"
+    And I fill in "Default front page" with "/test-page"
+    And I press "Save configuration"
+    When I am on "/"
+    And I click "français"
+    Then the url should match "/test-page_fr"
```
Comment (hernani) [inline tests/features/language-negotiation-suffix.feature]: We should also test that landing in /_fr redirects to the correct page.

File: modules/oe_multilingual_url_suffix/tests/src/Unit/LanguageNegotiationUrlSuffixTest.php [hunk 6]
```diff
@@ -102,7 +112,7 @@ public function testPathSuffix(string $suffix, array $suffixes, string $expected
     ]);
 
     $request = Request::create('/foo_' . $suffix, 'GET');
-    $method = new LanguageNegotiationUrlSuffix();
+    $method = new LanguageNegotiationUrlSuffix($this->aliasManager);
```
Comment (upchuk) [inline modules/oe_multilingual_url_suffix/tests/src/Unit/LanguageNegotiationUrlSuffixTest.php]: Why is this needed? The `LanguageNegotiationUrlSuffix` has no dependency on the alias manager nor even a constructor

File: tests/features/language-negotiation-suffix.feature [hunk 7]
```diff
@@ -53,3 +53,14 @@ Feature: Language negotiation by suffix.
     Then I should see the error message 'The suffix for Spanish, es, is not unique.'
     Then I should see the error message 'The suffix for French, es, is not unique.'
 
+  @front-page
+  Scenario: Language switcher should maintain front page alias in the links.
```
Comment (upchuk) [inline tests/features/language-negotiation-suffix.feature]: I'm guessing this should go in a separate behat feature since it's a completely separate module?

File: modules/oe_multilingual_front_page/README.md [hunk 8]
```diff
@@ -0,0 +1,7 @@
+# OpenEuropa Multilingual Front page
+
+The module ensures that the homepage alias is always maintained when linking to the front page.
+
+Note: Due to the nature of the suffix negotiation the front page alias is enforced 
+in order to avoid urls like ("/_en", "/_fr"). So links that are pointing
+to the front page will most likely contain the configured front page alias.
```
Comment (upchuk) [inline modules/oe_multilingual_front_page/README.md]: `most likely` ?

File: composer.json [hunk 9]
```diff
@@ -6,13 +6,16 @@
     "minimum-stability": "dev",
     "prefer-stable": true,
     "require": {
+        "php": "^7.1",
         "drupal/core": "^8.7",
         "drupal/language_selection_page": "^2.2",
         "drupal/pathauto": "^1.2",
-        "php": "^7.1"
+        "drupal/redirect": "^1.4"
```
Comment (upchuk) [inline composer.json]: Hm...the Redirect module is not required by oe_multilingual. Only by the submodule. So we need to require it as a dev dependency and add this information to the readme. See the poetry submodule wich requires a library as an example: https://github.com/openeuropa/oe_translation/tree/POETRY/modules/oe_translation_poetry

File: modules/oe_multilingual_front_page/oe_multilingual_front_page.info.yml [hunk 10]
```diff
@@ -0,0 +1,9 @@
+name: 'OpenEuropa Multilingual Front page'
+type: module
+description: 'Provides handling for front page.'
+core: 8.x
+package: 'OpenEuropa'
+
+dependencies:
+  - oe_multilingual:oe_multilingual
```
Comment (upchuk) [inline modules/oe_multilingual_front_page/oe_multilingual_front_page.info.yml]: Does it actually require oe_multingual?

File: modules/oe_multilingual_front_page/tests/src/Kernel/PathProcessorFrontPageTest.php [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+  }
+
+  use ContentTypeCreationTrait {
+    createContentType as drupalCreateContentType;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static $modules = [
```
Comment (upchuk) [inline modules/oe_multilingual_front_page/tests/src/Kernel/PathProcessorFrontPageTest.php]: The parent is `protected`. So if we override, we need to keep it the same.

File: modules/oe_multilingual_front_page/tests/src/Kernel/PathProcessorFrontPageTest.php [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+   *
+   * @covers ::processOutbound
+   */
+  public function testFrontPagePath() {
+    $this->drupalCreateContentType(['type' => 'article', 'name' => 'Article']);
+    $node = $this->drupalCreateNode(['type' => 'article', 'title' => 'Test page']);
+    // Set the node as front page.
+    $this->config('system.site')->set('page.front', '/node/' . $node->id())->save();
+    // Set node alias.
+    \Drupal::service('path.alias_storage')->save($node->toUrl()->toString(), '/test-page', LanguageInterface::LANGCODE_NOT_SPECIFIED);
```
Comment (upchuk) [inline modules/oe_multilingual_front_page/tests/src/Kernel/PathProcessorFrontPageTest.php]: You can putt he alias storage into a variable at the top and reuse it multiple times in the method.

File: modules/oe_multilingual_front_page/tests/src/Kernel/PathProcessorFrontPageTest.php [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+    $node_alias = \Drupal::service('path.alias_manager')->getAliasByPath('/node/' . $node->id());
```
Comment (upchuk) [inline modules/oe_multilingual_front_page/tests/src/Kernel/PathProcessorFrontPageTest.php]: Why do you keep loading the node_alias so many times?

File: modules/oe_multilingual_front_page/tests/src/Kernel/PathProcessorFrontPageTest.php [hunk 14]
```diff
@@ -0,0 +1,119 @@
...
+    $front_uri = \Drupal::config('system.site')->get('page.front');
+    $front_alias = \Drupal::service('path.alias_manager')->getAliasByPath($front_uri);
+    $this->assertEquals($front_alias, '/new-alias');
+
+    // Remove node alias.
+    \Drupal::service('path.alias_storage')->delete(['alias' => $node_alias]);
+    $node_alias = \Drupal::service('path.alias_manager')->getAliasByPath('/node/' . $node->id());
+    $front_uri = \Drupal::config('system.site')->get('page.front');
+    $front_alias = \Drupal::service('path.alias_manager')->getAliasByPath($front_uri);
+    $this->assertEquals($node_alias, $front_alias);
```
Comment (upchuk) [inline modules/oe_multilingual_front_page/tests/src/Kernel/PathProcessorFrontPageTest.php]: It's unclear what we are asserting here after deleting the node alias. Also, please respect the correct order of the assert method parameters (expected and actual)

File: modules/oe_multilingual_front_page/tests/src/Kernel/PathProcessorFrontPageTest.php [hunk 15]
```diff
@@ -0,0 +1,119 @@
...
+
+    // Set a different front page.
+    $this->config('system.site')->set('page.front', '/user')->save();
+    $front_uri = \Drupal::config('system.site')->get('page.front');
+    $front_alias = \Drupal::service('path.alias_manager')->getAliasByPath($front_uri);
+    $this->assertNotEqual($front_alias, $node_alias);
```
Comment (upchuk) [inline modules/oe_multilingual_front_page/tests/src/Kernel/PathProcessorFrontPageTest.php]: What are we asserting here?

File: modules/oe_multilingual_front_page/tests/src/Kernel/PathProcessorFrontPageTest.php [hunk 16]
```diff
@@ -0,0 +1,119 @@
...
+    // New node translatable node.
+    $node = $this->drupalCreateNode(['type' => 'oe_demo_translatable_page', 'title' => 'Translatable page']);
+    $node->addTranslation('fr', ['title' => 'Translatable page fr']);
+    \Drupal::service('path.alias_storage')->save($node->toUrl()->toString(), '/translatable-page', LanguageInterface::LANGCODE_NOT_SPECIFIED);
+    $this->config('system.site')->set('page.front', '/node/' . $node->id())->save();
+    $front_uri = \Drupal::config('system.site')->get('page.front');
+    $front_alias = \Drupal::service('path.alias_manager')->getAliasByPath($front_uri);
+    $this->assertEquals('/translatable-page', $front_alias);
+    \Drupal::configFactory()->getEditable('system.site')->set('default_langcode', 'fr')->save();
+    $this->assertEquals('/translatable-page', $front_alias);
```
Comment (upchuk) [inline modules/oe_multilingual_front_page/tests/src/Kernel/PathProcessorFrontPageTest.php]: Please add a comment as to why it's the same alias

File: modules/oe_multilingual_url_suffix/tests/src/Unit/LanguageNegotiationUrlSuffixTest.php [hunk 17]
```diff
@@ -24,6 +24,13 @@ class LanguageNegotiationUrlSuffixTest extends UnitTestCase {
    */
   protected $languageManager;
 
+  /**
```
Comment (upchuk) [inline modules/oe_multilingual_url_suffix/tests/src/Unit/LanguageNegotiationUrlSuffixTest.php]: There should be no change in this file.

File: tests/features/language-negotiation-suffix.feature [hunk 18]
```diff
@@ -52,4 +52,3 @@ Feature: Language negotiation by suffix.
     And I press "Save configuration"
     Then I should see the error message 'The suffix for Spanish, es, is not unique.'
     Then I should see the error message 'The suffix for French, es, is not unique.'
-
```
Comment (upchuk) [inline tests/features/language-negotiation-suffix.feature]: There should be no change in this file.

File: modules/oe_multilingual_front_page/tests/src/Kernel/PathProcessorFrontPageTest.php [hunk 19]
```diff
@@ -0,0 +1,119 @@
...
+
+    // Set a different front page.
+    $this->config('system.site')->set('page.front', '/user')->save();
+    $front_uri = \Drupal::config('system.site')->get('page.front');
+    $front_alias = \Drupal::service('path.alias_manager')->getAliasByPath($front_uri);
+    $this->assertNotEqual($front_alias, $node_alias);
+
+    // New node translatable node.
+    $node = $this->drupalCreateNode(['type' => 'oe_demo_translatable_page', 'title' => 'Translatable page']);
+    $node->addTranslation('fr', ['title' => 'Translatable page fr']);
```
Comment (upchuk) [inline modules/oe_multilingual_front_page/tests/src/Kernel/PathProcessorFrontPageTest.php]: Save? :)

File: modules/oe_multilingual_front_page/tests/src/Kernel/PathProcessorFrontPageTest.php [hunk 20]
```diff
@@ -0,0 +1,142 @@
...
+   * @var \Drupal\Core\Config\Config
+   */
+  protected $config;
+
+  /**
+   * The alias storage service.
+   *
+   * @var \Drupal\Core\Path\AliasStorageInterface
+   */
+  protected $aliasStorage;
```
Comment (upchuk) [inline modules/oe_multilingual_front_page/tests/src/Kernel/PathProcessorFrontPageTest.php]: This is not getting initialized. Why is it here?

File: modules/oe_multilingual_front_page/tests/src/Kernel/PathProcessorFrontPageTest.php [hunk 21]
```diff
@@ -0,0 +1,142 @@
...
+
+  /**
+   * The alias manager.
+   *
+   * @var \Drupal\Core\Path\AliasManager
+   */
+  protected $aliasManager;
```
Comment (upchuk) [inline modules/oe_multilingual_front_page/tests/src/Kernel/PathProcessorFrontPageTest.php]: This is not getting initialized. Why is it here?

File: modules/oe_multilingual_front_page/tests/src/Kernel/PathProcessorFrontPageTest.php [hunk 22]
```diff
@@ -0,0 +1,142 @@
...
+    'token',
+    'user',
+  ];
+
+  /**
+   * A config for retrieving required config settings.
+   *
+   * @var \Drupal\Core\Config\Config
+   */
+  protected $config;
```
Comment (upchuk) [inline modules/oe_multilingual_front_page/tests/src/Kernel/PathProcessorFrontPageTest.php]: This is not getting initialized. Why is it here?

File: modules/oe_multilingual_front_page/tests/src/Kernel/PathProcessorFrontPageTest.php [hunk 23]
```diff
@@ -0,0 +1,142 @@
...
+    // Set the node as front page.
+    $this->config('system.site')->set('page.front', '/node/' . $node->id())->save();
+    // Set node alias.
+    $alias_storage->save($node->toUrl()->toString(), '/test-page', LanguageInterface::LANGCODE_NOT_SPECIFIED);
+    $node_alias = $alias_manager->getAliasByPath('/node/' . $node->id());
+    $front_uri = \Drupal::config('system.site')->get('page.front');
+    $front_alias = $alias_manager->getAliasByPath($front_uri);
+    $url = Url::fromRoute('<front>')->toString();
+    $this->assertEquals($front_alias, '/test-page');
+    $this->assertEquals($front_alias, $url);
```
Comment (upchuk) [inline modules/oe_multilingual_front_page/tests/src/Kernel/PathProcessorFrontPageTest.php]: Just like I mentioned in previous comments, we need to assert actual values and not two variables that can have whatever values behind. We are never sure of the tests.

You set the homepage to Node X, and for node X you set the alias to `/test-page`. So if you make a Url for `<front>`, what do you expect the alias for that to be? 

Please adjust also below..there are other cases as well.

File: modules/oe_multilingual_front_page/tests/src/Kernel/PathProcessorFrontPageTest.php [hunk 24]
```diff
@@ -0,0 +1,142 @@
...
+    $alias_manager = $this->container->get('path.alias_manager');
+
+    $this->drupalCreateContentType(['type' => 'article', 'name' => 'Article']);
+    $node = $this->drupalCreateNode(['type' => 'article', 'title' => 'Test page']);
+    // Set the node as front page.
+    $this->config('system.site')->set('page.front', '/node/' . $node->id())->save();
+    // Set node alias.
+    $alias_storage->save($node->toUrl()->toString(), '/test-page', LanguageInterface::LANGCODE_NOT_SPECIFIED);
+    $node_alias = $alias_manager->getAliasByPath('/node/' . $node->id());
+    $front_uri = \Drupal::config('system.site')->get('page.front');
```
Comment (upchuk) [inline modules/oe_multilingual_front_page/tests/src/Kernel/PathProcessorFrontPageTest.php]: Just above you used the parent's class `config()` to get the config factory. Why you using `\Drupal::config()` here?

File: modules/oe_multilingual_front_page/tests/src/Kernel/PathProcessorFrontPageTest.php [hunk 25]
```diff
@@ -0,0 +1,142 @@
...
+    $front_alias = $alias_manager->getAliasByPath($front_uri);
+    $url = Url::fromRoute('<front>')->toString();
+    $this->assertEquals($front_alias, '/test-page');
+    $this->assertEquals($front_alias, $url);
+
+    // Update node alias.
+    $node_alias = \Drupal::service('path.alias_storage')->load(['alias' => $node_alias]);
```
Comment (upchuk) [inline modules/oe_multilingual_front_page/tests/src/Kernel/PathProcessorFrontPageTest.php]: Don't you have already the alias storage loaded above?

File: modules/oe_multilingual_front_page/tests/src/Kernel/PathProcessorFrontPageTest.php [hunk 26]
```diff
@@ -0,0 +1,142 @@
...
+    $alias_storage->save($front_uri, '/new-alias', LanguageInterface::LANGCODE_NOT_SPECIFIED, $node_alias['pid']);
+    $front_uri = \Drupal::config('system.site')->get('page.front');
```
Comment (upchuk) [inline modules/oe_multilingual_front_page/tests/src/Kernel/PathProcessorFrontPageTest.php]: Same comment

File: modules/oe_multilingual_front_page/tests/src/Kernel/PathProcessorFrontPageTest.php [hunk 27]
```diff
@@ -0,0 +1,142 @@
...
+    $front_alias = $alias_manager->getAliasByPath($front_uri);
+    // Check that the front page alias updates.
+    $this->assertEquals($front_alias, '/new-alias');
+    $url = Url::fromRoute('<front>')->toString();
+    $this->assertEquals($front_alias, $url);
+
+    // Remove node alias.
+    $alias_storage->delete($node_alias);
+    $node_alias = $alias_manager->getAliasByPath('/node/' . $node->id());
+    $front_uri = \Drupal::config('system.site')->get('page.front');
```
Comment (upchuk) [inline modules/oe_multilingual_front_page/tests/src/Kernel/PathProcessorFrontPageTest.php]: Same comment

File: modules/oe_multilingual_front_page/tests/src/Kernel/PathProcessorFrontPageTest.php [hunk 28]
```diff
@@ -0,0 +1,142 @@
...
+    $this->assertEquals($front_alias, $node_alias);
+    $url = Url::fromRoute('<front>')->toString();
+    $this->assertEquals($front_alias, $url);
+
+    // New node translatable node.
+    $node = $this->drupalCreateNode(['type' => 'oe_demo_translatable_page', 'title' => 'Translatable page']);
+    $node->addTranslation('fr', ['title' => 'Translatable page fr'])->save();
+    $alias_storage->save($node->toUrl()->toString(), '/translatable-page', LanguageInterface::LANGCODE_NOT_SPECIFIED);
+    $this->config('system.site')->set('page.front', '/node/' . $node->id())->save();
+    $front_uri = \Drupal::config('system.site')->get('page.front');
```
Comment (upchuk) [inline modules/oe_multilingual_front_page/tests/src/Kernel/PathProcessorFrontPageTest.php]: Same comment

File: modules/oe_multilingual_front_page/tests/src/Kernel/PathProcessorFrontPageTest.php [hunk 29]
```diff
@@ -0,0 +1,142 @@
...
+    $front_alias = $alias_manager->getAliasByPath($front_uri);
+    $this->assertEquals('/translatable-page', $front_alias);
+    $url = Url::fromRoute('<front>')->toString();
+    $this->assertEquals($front_alias, $url);
+    // Set the default language to French.
+    \Drupal::configFactory()->getEditable('system.site')->set('default_langcode', 'fr')->save();
```
Comment (upchuk) [inline modules/oe_multilingual_front_page/tests/src/Kernel/PathProcessorFrontPageTest.php]: Same comment. Here even in a third way.

File: modules/oe_multilingual_front_page/tests/src/Kernel/PathProcessorFrontPageTest.php [hunk 30]
```diff
@@ -0,0 +1,142 @@
...
+    $url = Url::fromRoute('<front>')->toString();
+    $this->assertEquals($front_alias, $url);
+
+    // Remove node alias.
+    $alias_storage->delete($node_alias);
+    $node_alias = $alias_manager->getAliasByPath('/node/' . $node->id());
+    $front_uri = \Drupal::config('system.site')->get('page.front');
+    $front_alias = $alias_manager->getAliasByPath($front_uri);
+    // Check that the front page alias updates.
+    $this->assertEquals($front_alias, $node_alias);
```
Comment (upchuk) [inline modules/oe_multilingual_front_page/tests/src/Kernel/PathProcessorFrontPageTest.php]: If you delete the alias of the node, you should assert that the front alias is `node/1` no? Same below after the URL generation.

File: modules/oe_multilingual_front_page/tests/src/Kernel/PathProcessorFrontPageTest.php (hunk 14 above)
Comment (upchuk) [reply modules/oe_multilingual_front_page/tests/src/Kernel/PathProcessorFrontPageTest.php]: Not addressed

File: modules/oe_multilingual_front_page/tests/src/Kernel/PathProcessorFrontPageTest.php (hunk 16 above)
Comment (upchuk) [reply modules/oe_multilingual_front_page/tests/src/Kernel/PathProcessorFrontPageTest.php]: Not addressed

File: tests/features/language-negotiation-suffix.feature (hunk 18 above)
Comment (upchuk) [reply tests/features/language-negotiation-suffix.feature]: Not addressed

File: modules/oe_multilingual_url_suffix/tests/src/Unit/LanguageNegotiationUrlSuffixTest.php [hunk 31]
```diff
@@ -42,7 +42,6 @@ class LanguageNegotiationUrlSuffixTest extends UnitTestCase {
    * {@inheritdoc}
    */
   protected function setUp() {
-
```
Comment (upchuk) [inline modules/oe_multilingual_url_suffix/tests/src/Unit/LanguageNegotiationUrlSuffixTest.php]: Why are we making changes in this file?
